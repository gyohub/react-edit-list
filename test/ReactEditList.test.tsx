import React from 'react';
import {cleanup, fireEvent, render, waitFor} from '@testing-library/react';

import ReactEditList, * as REL from 'react-edit-list';

import {schema, data, onLoad} from './data';

describe('<ReactEditList>', () => {
    it('basic use', async () => {
        const onLoadFn = jest.fn(onLoad);
        const r = render(<ReactEditList schema={schema} onLoad={onLoadFn} />);
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(onLoadFn).toBeCalledTimes(1);
        expect(r.container.innerHTML).toMatchSnapshot();

        r.rerender(
            <ReactEditList
                schema={schema}
                onLoad={onLoadFn}
                className='table table-light table-fixed align-middle'
                headClassName='table-dark'
                inputClassName='w-100'
                thClassName={{
                    product: 'col-4',
                    type: 'col-3',
                    price: 'col-2',
                    stock: 'col-2',
                    buttons: 'col-1'
                }}
                btnValidateClassName='btn btn-success p-0 m-0'
                btnDeleteClassName='btn btn-danger py-0 px-1 m-0 mx-1'
                btnCancelClassName='btn btn-secondary py-0 px-1 m-0 mx-1'
            />
        );
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(onLoadFn).toBeCalledTimes(2);
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    describe('onInsert()', () => {
        it('onInsert() undefined', async () => {
            let recvItem: REL.Row = {};
            const onInsert = jest.fn((item: REL.Row): void => {
                recvItem = item;
            });
            const r = render(<ReactEditList schema={schema} onLoad={onLoad} onInsert={onInsert} />);
            await waitFor(() => expect(r.getByText(/Desk/)));
            expect(r.container.innerHTML).toMatchSnapshot();

            const cells = r.container.querySelectorAll('td');
            fireEvent(cells[cells.length - 5], new MouseEvent('click', {bubbles: true}));
            await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
            expect(r.container.innerHTML).toMatchSnapshot();

            const inputs = r.container.querySelectorAll('input');
            fireEvent.change(inputs[0], {target: {value: 'test item'}});
            fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));
            expect(onInsert).toBeCalledTimes(1);

            await waitFor(() => expect(recvItem.product).toBe('test item'));
            expect(recvItem.type).toBeUndefined();
            expect(recvItem.price).toBeUndefined();
            expect(recvItem.stock).toBeUndefined();
            await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
            expect(r.container.innerHTML).toMatchSnapshot();
            r.unmount();
        });

        it('onInsert() false', async () => {
            let recvItem: REL.Row = {};
            const onInsert = jest.fn((item: REL.Row): Promise<false> => {
                recvItem = item;
                return Promise.resolve(false);
            });
            const r = render(<ReactEditList schema={schema} onLoad={onLoad} onInsert={onInsert} />);
            await waitFor(() => expect(r.getByText(/Desk/)));
            expect(r.container.innerHTML).toMatchSnapshot();

            const cells = r.container.querySelectorAll('td');
            fireEvent(cells[cells.length - 5], new MouseEvent('click', {bubbles: true}));
            await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
            expect(r.container.innerHTML).toMatchSnapshot();

            const inputs = r.container.querySelectorAll('input');
            fireEvent.change(inputs[0], {target: {value: 'test item 2'}});
            fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));
            expect(onInsert).toBeCalledTimes(1);

            await waitFor(() => expect(recvItem.product).toBe('test item 2'));
            expect(recvItem.type).toBeUndefined();
            expect(recvItem.price).toBeUndefined();
            expect(recvItem.stock).toBeUndefined();
            await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(3));
            expect(r.container.innerHTML).toMatchSnapshot();
            r.unmount();
        });

        it('onInsert() false', async () => {
            let recvItem: REL.Row = {};
            const onInsert = jest.fn((item: REL.Row): Promise<REL.Row> => {
                recvItem = item;
                return Promise.resolve({product: 'something different'});
            });
            const r = render(<ReactEditList schema={schema} onLoad={onLoad} onInsert={onInsert} />);
            await waitFor(() => expect(r.getByText(/Desk/)));
            expect(r.container.innerHTML).toMatchSnapshot();

            const cells = r.container.querySelectorAll('td');
            fireEvent(cells[cells.length - 5], new MouseEvent('click', {bubbles: true}));
            await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
            expect(r.container.innerHTML).toMatchSnapshot();

            const inputs = r.container.querySelectorAll('input');
            fireEvent.change(inputs[0], {target: {value: 'test item 3'}});
            fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));
            expect(onInsert).toBeCalledTimes(1);

            await waitFor(() => expect(recvItem.product).toBe('test item 3'));
            expect(recvItem.type).toBeUndefined();
            expect(recvItem.price).toBeUndefined();
            expect(recvItem.stock).toBeUndefined();
            await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
            await waitFor(() => expect(r.getByText(/something different/)));
            expect(r.container.innerHTML).toMatchSnapshot();
            r.unmount();
        });
    });
});
