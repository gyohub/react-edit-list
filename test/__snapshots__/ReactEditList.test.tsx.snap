// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReactEditList> basic use 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`<ReactEditList> basic use 2`] = `"<table class=\\"table table-light table-fixed align-middle\\"><thead class=\\"table-dark\\"><tr><th class=\\"col-4\\">product</th><th class=\\"col-3\\">type</th><th class=\\"col-2\\">price</th><th class=\\"col-2\\">stock</th><th class=\\"col-1\\"></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() false 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td>test item</td><td></td><td></td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() false 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td>test item</td><td></td><td></td><td></td><td><button>x</button></td></tr><tr><td><input type=\\"text\\" value=\\"\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"\\"></td><td><input type=\\"number\\" value=\\"\\"></td><td><button>✓</button><button>x</button></td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() false 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td>test item</td><td></td><td></td><td></td><td><button>x</button></td></tr><tr><td><input type=\\"text\\" value=\\"test item 2\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"\\"></td><td><input type=\\"number\\" value=\\"\\"></td><td><button>✓</button><button>x</button></td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() false 4`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td>test item</td><td></td><td></td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() false 5`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td>test item</td><td></td><td></td><td></td><td><button>x</button></td></tr><tr><td><input type=\\"text\\" value=\\"\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"\\"></td><td><input type=\\"number\\" value=\\"\\"></td><td><button>✓</button><button>x</button></td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() false 6`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td>test item</td><td></td><td></td><td></td><td><button>x</button></td></tr><tr><td>something different</td><td></td><td></td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() undefined 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() undefined 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td><input type=\\"text\\" value=\\"\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"\\"></td><td><input type=\\"number\\" value=\\"\\"></td><td><button>✓</button><button>x</button></td></tr></tbody></table>"`;

exports[`<ReactEditList> onInsert() onInsert() undefined 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td>test item</td><td></td><td></td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;
